{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a7847299",
   "metadata": {
    "_cell_guid": "b1076dfc-b9ad-4769-8c92-a6c4dae69d19",
    "_uuid": "8f2839f25d086af736a60e9eeb907d3b93b6e0e5",
    "execution": {
     "iopub.execute_input": "2022-07-29T10:47:31.935714Z",
     "iopub.status.busy": "2022-07-29T10:47:31.935216Z",
     "iopub.status.idle": "2022-07-29T10:47:31.950267Z",
     "shell.execute_reply": "2022-07-29T10:47:31.949389Z"
    },
    "papermill": {
     "duration": 0.023868,
     "end_time": "2022-07-29T10:47:31.953466",
     "exception": false,
     "start_time": "2022-07-29T10:47:31.929598",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/kaggle/input/titanic/train.csv\n",
      "/kaggle/input/titanic/test.csv\n",
      "/kaggle/input/titanic/gender_submission.csv\n"
     ]
    }
   ],
   "source": [
    "# This Python 3 environment comes with many helpful analytics libraries installed\n",
    "# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n",
    "# For example, here's several helpful packages to load\n",
    "\n",
    "import numpy as np # linear algebra\n",
    "import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n",
    "\n",
    "# Input data files are available in the read-only \"../input/\" directory\n",
    "# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n",
    "\n",
    "import os\n",
    "for dirname, _, filenames in os.walk('/kaggle/input'):\n",
    "    for filename in filenames:\n",
    "        print(os.path.join(dirname, filename))\n",
    "\n",
    "# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n",
    "# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df996750",
   "metadata": {
    "papermill": {
     "duration": 0.002703,
     "end_time": "2022-07-29T10:47:31.960348",
     "exception": false,
     "start_time": "2022-07-29T10:47:31.957645",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Get the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9bc16315",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-07-29T10:47:31.968136Z",
     "iopub.status.busy": "2022-07-29T10:47:31.967418Z",
     "iopub.status.idle": "2022-07-29T10:47:33.185159Z",
     "shell.execute_reply": "2022-07-29T10:47:33.183986Z"
    },
    "papermill": {
     "duration": 1.224745,
     "end_time": "2022-07-29T10:47:33.187951",
     "exception": false,
     "start_time": "2022-07-29T10:47:31.963206",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',\n",
       "       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Titanic competition in Kaggle\n",
    "## data downloaded from Kaggle.\n",
    "import pandas\n",
    "from sklearn.model_selection import train_test_split\n",
    "import numpy as np\n",
    "\n",
    "train = pandas.read_csv('/kaggle/input/titanic/train.csv')\n",
    "train.columns  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1f8017f7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-07-29T10:47:33.196928Z",
     "iopub.status.busy": "2022-07-29T10:47:33.196202Z",
     "iopub.status.idle": "2022-07-29T12:17:12.879741Z",
     "shell.execute_reply": "2022-07-29T12:17:12.878413Z"
    },
    "papermill": {
     "duration": 5379.695854,
     "end_time": "2022-07-29T12:17:12.887054",
     "exception": false,
     "start_time": "2022-07-29T10:47:33.191200",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.92      0.98      0.95       407\n",
      "           1       0.96      0.87      0.92       261\n",
      "\n",
      "    accuracy                           0.94       668\n",
      "   macro avg       0.94      0.93      0.93       668\n",
      "weighted avg       0.94      0.94      0.94       668\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       1.00      1.00      1.00       407\n",
      "           1       1.00      1.00      1.00       261\n",
      "\n",
      "    accuracy                           1.00       668\n",
      "   macro avg       1.00      1.00      1.00       668\n",
      "weighted avg       1.00      1.00      1.00       668\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:372: FitFailedWarning: \n",
      "3 fits failed out of a total of 30.\n",
      "The score on these train-test partitions for these parameters will be set to nan.\n",
      "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
      "\n",
      "Below are more details about the failures:\n",
      "--------------------------------------------------------------------------------\n",
      "3 fits failed with the following error:\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py\", line 680, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/pipeline.py\", line 394, in fit\n",
      "    self._final_estimator.fit(Xt, y, **fit_params_last_step)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py\", line 1541, in fit\n",
      "    sample_weight=sample_weight,\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/svm/_base.py\", line 1198, in _fit_liblinear\n",
      "    sample_weight,\n",
      "  File \"sklearn/svm/_liblinear.pyx\", line 52, in sklearn.svm._liblinear.train_wrap\n",
      "ValueError: b'C <= 0'\n",
      "\n",
      "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_search.py:972: UserWarning: One or more of the test scores are non-finite: [0.84028192 0.84027363 0.83694859 0.83528192 0.84194859 0.84527363\n",
      " 0.83194859        nan 0.83194859 0.83863184]\n",
      "  category=UserWarning,\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:372: FitFailedWarning: \n",
      "6 fits failed out of a total of 30.\n",
      "The score on these train-test partitions for these parameters will be set to nan.\n",
      "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
      "\n",
      "Below are more details about the failures:\n",
      "--------------------------------------------------------------------------------\n",
      "6 fits failed with the following error:\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py\", line 680, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/pipeline.py\", line 394, in fit\n",
      "    self._final_estimator.fit(Xt, y, **fit_params_last_step)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py\", line 1541, in fit\n",
      "    sample_weight=sample_weight,\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/svm/_base.py\", line 1198, in _fit_liblinear\n",
      "    sample_weight,\n",
      "  File \"sklearn/svm/_liblinear.pyx\", line 52, in sklearn.svm._liblinear.train_wrap\n",
      "ValueError: b'C <= 0'\n",
      "\n",
      "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_search.py:972: UserWarning: One or more of the test scores are non-finite: [0.82029022 0.83198176 0.83194859        nan 0.8303068  0.83194859\n",
      " 0.82864013        nan 0.83029851 0.84194859]\n",
      "  category=UserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       1.00      1.00      1.00       407\n",
      "           1       1.00      1.00      1.00       261\n",
      "\n",
      "    accuracy                           1.00       668\n",
      "   macro avg       1.00      1.00      1.00       668\n",
      "weighted avg       1.00      1.00      1.00       668\n",
      "\n",
      "0.8487109905020352\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.88      0.86      0.87       142\n",
      "           1       0.76      0.79      0.78        81\n",
      "\n",
      "    accuracy                           0.83       223\n",
      "   macro avg       0.82      0.82      0.82       223\n",
      "weighted avg       0.84      0.83      0.83       223\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.98      0.99      0.99       407\n",
      "           1       0.98      0.97      0.98       261\n",
      "\n",
      "    accuracy                           0.98       668\n",
      "   macro avg       0.98      0.98      0.98       668\n",
      "weighted avg       0.98      0.98      0.98       668\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       1.00      1.00      1.00       407\n",
      "           1       0.99      1.00      0.99       261\n",
      "\n",
      "    accuracy                           1.00       668\n",
      "   macro avg       0.99      1.00      1.00       668\n",
      "weighted avg       1.00      1.00      1.00       668\n",
      "\n",
      "0.8382673100662104\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.88      0.86      0.87       142\n",
      "           1       0.76      0.80      0.78        81\n",
      "\n",
      "    accuracy                           0.84       223\n",
      "   macro avg       0.82      0.83      0.83       223\n",
      "weighted avg       0.84      0.84      0.84       223\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       1.00      1.00      1.00       407\n",
      "           1       1.00      1.00      1.00       261\n",
      "\n",
      "    accuracy                           1.00       668\n",
      "   macro avg       1.00      1.00      1.00       668\n",
      "weighted avg       1.00      1.00      1.00       668\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.85      0.90      0.88       142\n",
      "           1       0.81      0.73      0.77        81\n",
      "\n",
      "    accuracy                           0.84       223\n",
      "   macro avg       0.83      0.81      0.82       223\n",
      "weighted avg       0.84      0.84      0.84       223\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.85      0.97      0.91       407\n",
      "           1       0.95      0.74      0.83       261\n",
      "\n",
      "    accuracy                           0.88       668\n",
      "   macro avg       0.90      0.86      0.87       668\n",
      "weighted avg       0.89      0.88      0.88       668\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.80      0.91      0.85       142\n",
      "           1       0.79      0.60      0.69        81\n",
      "\n",
      "    accuracy                           0.80       223\n",
      "   macro avg       0.80      0.76      0.77       223\n",
      "weighted avg       0.80      0.80      0.79       223\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.73      0.88      0.80       407\n",
      "           1       0.72      0.49      0.58       261\n",
      "\n",
      "    accuracy                           0.73       668\n",
      "   macro avg       0.72      0.68      0.69       668\n",
      "weighted avg       0.72      0.73      0.71       668\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       1.00      1.00      1.00       407\n",
      "           1       1.00      1.00      1.00       261\n",
      "\n",
      "    accuracy                           1.00       668\n",
      "   macro avg       1.00      1.00      1.00       668\n",
      "weighted avg       1.00      1.00      1.00       668\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.63      0.85      0.72       142\n",
      "           1       0.33      0.14      0.19        81\n",
      "\n",
      "    accuracy                           0.59       223\n",
      "   macro avg       0.48      0.49      0.46       223\n",
      "weighted avg       0.52      0.59      0.53       223\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:372: FitFailedWarning: \n",
      "3 fits failed out of a total of 30.\n",
      "The score on these train-test partitions for these parameters will be set to nan.\n",
      "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
      "\n",
      "Below are more details about the failures:\n",
      "--------------------------------------------------------------------------------\n",
      "3 fits failed with the following error:\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py\", line 680, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/pipeline.py\", line 394, in fit\n",
      "    self._final_estimator.fit(Xt, y, **fit_params_last_step)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py\", line 1541, in fit\n",
      "    sample_weight=sample_weight,\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/svm/_base.py\", line 1198, in _fit_liblinear\n",
      "    sample_weight,\n",
      "  File \"sklearn/svm/_liblinear.pyx\", line 52, in sklearn.svm._liblinear.train_wrap\n",
      "ValueError: b'C <= 0'\n",
      "\n",
      "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_search.py:972: UserWarning: One or more of the test scores are non-finite: [0.83680093 0.84428823 0.83680766 0.83529269 0.83980393 0.8428002\n",
      " 0.83831589        nan 0.8398174  0.82630388]\n",
      "  category=UserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       1.00      1.00      1.00       407\n",
      "           1       1.00      1.00      1.00       261\n",
      "\n",
      "    accuracy                           1.00       668\n",
      "   macro avg       1.00      1.00      1.00       668\n",
      "weighted avg       1.00      1.00      1.00       668\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:372: FitFailedWarning: \n",
      "3 fits failed out of a total of 30.\n",
      "The score on these train-test partitions for these parameters will be set to nan.\n",
      "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
      "\n",
      "Below are more details about the failures:\n",
      "--------------------------------------------------------------------------------\n",
      "3 fits failed with the following error:\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py\", line 680, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/pipeline.py\", line 394, in fit\n",
      "    self._final_estimator.fit(Xt, y, **fit_params_last_step)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py\", line 1541, in fit\n",
      "    sample_weight=sample_weight,\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/svm/_base.py\", line 1198, in _fit_liblinear\n",
      "    sample_weight,\n",
      "  File \"sklearn/svm/_liblinear.pyx\", line 52, in sklearn.svm._liblinear.train_wrap\n",
      "ValueError: b'C <= 0'\n",
      "\n",
      "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_search.py:972: UserWarning: One or more of the test scores are non-finite: [0.82863184 0.8303068  0.83028192 0.83529022 0.83363184        nan\n",
      " 0.82864842 0.83529851 0.82863184 0.8303068 ]\n",
      "  category=UserWarning,\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:372: FitFailedWarning: \n",
      "3 fits failed out of a total of 30.\n",
      "The score on these train-test partitions for these parameters will be set to nan.\n",
      "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
      "\n",
      "Below are more details about the failures:\n",
      "--------------------------------------------------------------------------------\n",
      "3 fits failed with the following error:\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py\", line 680, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/pipeline.py\", line 394, in fit\n",
      "    self._final_estimator.fit(Xt, y, **fit_params_last_step)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py\", line 1541, in fit\n",
      "    sample_weight=sample_weight,\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/svm/_base.py\", line 1198, in _fit_liblinear\n",
      "    sample_weight,\n",
      "  File \"sklearn/svm/_liblinear.pyx\", line 52, in sklearn.svm._liblinear.train_wrap\n",
      "ValueError: b'C <= 0'\n",
      "\n",
      "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_search.py:972: UserWarning: One or more of the test scores are non-finite: [0.83529851 0.83529022        nan 0.83363184 0.83862355 0.83194859\n",
      " 0.83694859 0.84693201 0.83361526 0.8469403 ]\n",
      "  category=UserWarning,\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:372: FitFailedWarning: \n",
      "3 fits failed out of a total of 30.\n",
      "The score on these train-test partitions for these parameters will be set to nan.\n",
      "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
      "\n",
      "Below are more details about the failures:\n",
      "--------------------------------------------------------------------------------\n",
      "3 fits failed with the following error:\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py\", line 680, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/pipeline.py\", line 394, in fit\n",
      "    self._final_estimator.fit(Xt, y, **fit_params_last_step)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py\", line 1541, in fit\n",
      "    sample_weight=sample_weight,\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/svm/_base.py\", line 1198, in _fit_liblinear\n",
      "    sample_weight,\n",
      "  File \"sklearn/svm/_liblinear.pyx\", line 52, in sklearn.svm._liblinear.train_wrap\n",
      "ValueError: b'C <= 0'\n",
      "\n",
      "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_search.py:972: UserWarning: One or more of the test scores are non-finite: [0.81225539        nan 0.8305141  0.83218076 0.83053897 0.83218076\n",
      " 0.83218076 0.83385572 0.8355141  0.8421393 ]\n",
      "  category=UserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.85      0.92      0.89       142\n",
      "           1       0.84      0.72      0.77        81\n",
      "\n",
      "    accuracy                           0.85       223\n",
      "   macro avg       0.85      0.82      0.83       223\n",
      "weighted avg       0.85      0.85      0.84       223\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:372: FitFailedWarning: \n",
      "3 fits failed out of a total of 30.\n",
      "The score on these train-test partitions for these parameters will be set to nan.\n",
      "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
      "\n",
      "Below are more details about the failures:\n",
      "--------------------------------------------------------------------------------\n",
      "3 fits failed with the following error:\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py\", line 680, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/pipeline.py\", line 394, in fit\n",
      "    self._final_estimator.fit(Xt, y, **fit_params_last_step)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py\", line 1541, in fit\n",
      "    sample_weight=sample_weight,\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/svm/_base.py\", line 1198, in _fit_liblinear\n",
      "    sample_weight,\n",
      "  File \"sklearn/svm/_liblinear.pyx\", line 52, in sklearn.svm._liblinear.train_wrap\n",
      "ValueError: b'C <= 0'\n",
      "\n",
      "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_search.py:972: UserWarning: One or more of the test scores are non-finite: [0.84279347 0.8397972  0.8472845  0.84278673        nan 0.84130543\n",
      " 0.84278673 0.83830243 0.8412987  0.8398174 ]\n",
      "  category=UserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       1.00      1.00      1.00       407\n",
      "           1       1.00      1.00      1.00       261\n",
      "\n",
      "    accuracy                           1.00       668\n",
      "   macro avg       1.00      1.00      1.00       668\n",
      "weighted avg       1.00      1.00      1.00       668\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:372: FitFailedWarning: \n",
      "3 fits failed out of a total of 30.\n",
      "The score on these train-test partitions for these parameters will be set to nan.\n",
      "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
      "\n",
      "Below are more details about the failures:\n",
      "--------------------------------------------------------------------------------\n",
      "3 fits failed with the following error:\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py\", line 680, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/pipeline.py\", line 394, in fit\n",
      "    self._final_estimator.fit(Xt, y, **fit_params_last_step)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py\", line 1541, in fit\n",
      "    sample_weight=sample_weight,\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/svm/_base.py\", line 1198, in _fit_liblinear\n",
      "    sample_weight,\n",
      "  File \"sklearn/svm/_liblinear.pyx\", line 52, in sklearn.svm._liblinear.train_wrap\n",
      "ValueError: b'C <= 0'\n",
      "\n",
      "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_search.py:972: UserWarning: One or more of the test scores are non-finite: [0.8469403  0.8369403  0.83195688 0.83693201 0.84360697 0.84359867\n",
      " 0.84028192 0.84859867 0.84193201        nan]\n",
      "  category=UserWarning,\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:372: FitFailedWarning: \n",
      "3 fits failed out of a total of 30.\n",
      "The score on these train-test partitions for these parameters will be set to nan.\n",
      "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
      "\n",
      "Below are more details about the failures:\n",
      "--------------------------------------------------------------------------------\n",
      "3 fits failed with the following error:\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py\", line 680, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/pipeline.py\", line 394, in fit\n",
      "    self._final_estimator.fit(Xt, y, **fit_params_last_step)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py\", line 1541, in fit\n",
      "    sample_weight=sample_weight,\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/svm/_base.py\", line 1198, in _fit_liblinear\n",
      "    sample_weight,\n",
      "  File \"sklearn/svm/_liblinear.pyx\", line 52, in sklearn.svm._liblinear.train_wrap\n",
      "ValueError: b'C <= 0'\n",
      "\n",
      "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_search.py:972: UserWarning: One or more of the test scores are non-finite: [0.83029022 0.82862355 0.82695688 0.83361526 0.83361526 0.82862355\n",
      " 0.82695688 0.83194859        nan 0.82362355]\n",
      "  category=UserWarning,\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:372: FitFailedWarning: \n",
      "6 fits failed out of a total of 30.\n",
      "The score on these train-test partitions for these parameters will be set to nan.\n",
      "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
      "\n",
      "Below are more details about the failures:\n",
      "--------------------------------------------------------------------------------\n",
      "6 fits failed with the following error:\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py\", line 680, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/pipeline.py\", line 394, in fit\n",
      "    self._final_estimator.fit(Xt, y, **fit_params_last_step)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py\", line 1541, in fit\n",
      "    sample_weight=sample_weight,\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/svm/_base.py\", line 1198, in _fit_liblinear\n",
      "    sample_weight,\n",
      "  File \"sklearn/svm/_liblinear.pyx\", line 52, in sklearn.svm._liblinear.train_wrap\n",
      "ValueError: b'C <= 0'\n",
      "\n",
      "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_search.py:972: UserWarning: One or more of the test scores are non-finite: [0.8355058         nan 0.82389718        nan 0.82722222 0.82556385\n",
      " 0.83883085 0.81228856 0.83218905 0.82722222]\n",
      "  category=UserWarning,\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py:372: FitFailedWarning: \n",
      "3 fits failed out of a total of 30.\n",
      "The score on these train-test partitions for these parameters will be set to nan.\n",
      "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
      "\n",
      "Below are more details about the failures:\n",
      "--------------------------------------------------------------------------------\n",
      "3 fits failed with the following error:\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_validation.py\", line 680, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/pipeline.py\", line 394, in fit\n",
      "    self._final_estimator.fit(Xt, y, **fit_params_last_step)\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py\", line 1541, in fit\n",
      "    sample_weight=sample_weight,\n",
      "  File \"/opt/conda/lib/python3.7/site-packages/sklearn/svm/_base.py\", line 1198, in _fit_liblinear\n",
      "    sample_weight,\n",
      "  File \"sklearn/svm/_liblinear.pyx\", line 52, in sklearn.svm._liblinear.train_wrap\n",
      "ValueError: b'C <= 0'\n",
      "\n",
      "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
      "/opt/conda/lib/python3.7/site-packages/sklearn/model_selection/_search.py:972: UserWarning: One or more of the test scores are non-finite: [0.84715589 0.8405058  0.8338806  0.82389718 0.84052239 0.84384743\n",
      " 0.83554726 0.83886401 0.83886401        nan]\n",
      "  category=UserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.88      0.87      0.88       142\n",
      "           1       0.77      0.80      0.79        81\n",
      "\n",
      "    accuracy                           0.84       223\n",
      "   macro avg       0.83      0.83      0.83       223\n",
      "weighted avg       0.84      0.84      0.84       223\n",
      "\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',\n",
    "       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],\n",
    "      dtype='object')\n",
    "'''  \n",
    "surname = []\n",
    "title = []\n",
    "for txt in train['Name']:\n",
    "      txt1 = txt.split(',')[0]\n",
    "      txt2 = (txt.split(\",\")[1]).split(\".\")[0]\n",
    "      surname.append(txt1)\n",
    "      title.append(txt2)\n",
    "\n",
    "train['Surname'] = pandas.DataFrame(surname)\n",
    "train['Title'] = pandas.DataFrame(title, dtype='category')\n",
    "\n",
    "# There are too many missing values in \"Cabin\" with uncertain impact \n",
    "# to the accuracy after imputation. (It seems improve a little bit \n",
    "# after a comparison.)\n",
    "\n",
    "X = train.drop(['Survived','Name'],axis=1)\n",
    "y = (train.copy())['Survived']\n",
    "\n",
    "# Split the (training) data to train and test\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=4686)\n",
    "\n",
    "# missing values in X_train\n",
    "np.sum(X_train.isna(), axis=0)\n",
    "\n",
    "# Fix missing values with means of corresponding group.\n",
    "# age => Title and Sex\n",
    "# Embarked => Pclass and Sex.\n",
    "# Imputation after splitting, in order not to overfit. \n",
    "X_train[\"Age\"] = X_train[\"Age\"].fillna(X_train.groupby(['Sex','Title'])['Age'].transform('mean'))\n",
    "X_test[\"Age\"] = X_test[\"Age\"].fillna(X_test.groupby(['Sex','Title'])['Age'].transform('mean'))\n",
    "\n",
    "from sklearn.impute import SimpleImputer, KNNImputer\n",
    "from sklearn.preprocessing import OneHotEncoder, PolynomialFeatures\n",
    "from sklearn.pipeline import Pipeline\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.compose import ColumnTransformer\n",
    "from sklearn.metrics import classification_report\n",
    "from sklearn.model_selection import cross_val_score, RandomizedSearchCV, GridSearchCV\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "Impute1 = SimpleImputer(strategy='constant')\n",
    "Impute2 = SimpleImputer(strategy=\"most_frequent\")\n",
    "Encode = OneHotEncoder(handle_unknown=\"ignore\")\n",
    "Scale = StandardScaler()\n",
    "\n",
    "Embarked = ['Embarked']\n",
    "Cat_attribs = ['Pclass', 'Surname','Title', 'Sex', 'Ticket', 'Cabin']\n",
    "Age = ['Age']\n",
    "Num_attribs = ['Fare']\n",
    "\n",
    "num_pipiline1 = Pipeline(steps=[\n",
    "    ('scale1', StandardScaler())\n",
    "])\n",
    "\n",
    "num_pipiline2 = Pipeline(steps=[\n",
    "      ('poly', PolynomialFeatures()),\n",
    "    ('scale1', StandardScaler())\n",
    "])\n",
    "\n",
    "cat_pipeline1 = Pipeline(steps=[\n",
    "    ('impute2', Impute2),\n",
    "    ('enco', Encode)\n",
    "])\n",
    "\n",
    "cat_pipeline2 = Pipeline(steps=[ \n",
    "      ('impute1', Impute1),\n",
    "    ('encode', Encode)\n",
    "])\n",
    "\n",
    "total_pipeline = ColumnTransformer(transformers=[\n",
    "    ('num_imp', num_pipiline1, Age),\n",
    "    ('num_scale', num_pipiline2, Num_attribs),\n",
    "    ('cat_imp', cat_pipeline1, Embarked),\n",
    "    ('cat_encode', cat_pipeline2, Cat_attribs)\n",
    "], remainder=\"passthrough\")\n",
    "\n",
    "'''\n",
    "Logistic regression model.\n",
    "'''\n",
    "Logit = Pipeline(steps=[ \n",
    "    ('preprocess', total_pipeline),\n",
    "    ('logit', LogisticRegression(penalty='l2', C=1, max_iter=1000, solver=\"liblinear\") )\n",
    "]\n",
    ")\n",
    "\n",
    "titanic_logit = Logit.fit(X_train, y_train)\n",
    "titanic_pred = Logit.predict(X_train)\n",
    "print(classification_report(y_train, titanic_pred))\n",
    "# accuracy 0.94\n",
    "\n",
    "cross_logit = cross_val_score(Logit, X_train, y_train, cv=10)\n",
    "np.mean(cross_logit)\n",
    "# ~ 0.84\n",
    "\n",
    "dist = {\n",
    "      'preprocess__num_scale__poly__degree': range(2,4),\n",
    "    'logit__C': range(0,1000,50), \n",
    "    'logit__penalty': ['l2','l1']\n",
    "}\n",
    "\n",
    "lr_clf = RandomizedSearchCV(Logit, param_distributions=dist, n_iter=10, cv=3)\n",
    "\n",
    "lr_clf.fit(X_train, y_train)\n",
    "\n",
    "pred_lg2 = lr_clf.predict(X_train)\n",
    "print(classification_report(y_train, pred_lg2))\n",
    "# 1 !!\n",
    "\n",
    "cross_lg2 = cross_val_score(lr_clf, X_train, y_train, cv=10)\n",
    "np.mean(cross_lg2)\n",
    "# 0.842740841248304 Overfitting.\n",
    "\n",
    "Logit2 = Pipeline(steps=[ \n",
    "    ('preprocess', total_pipeline),\n",
    "    ('logit', LogisticRegression(penalty='l2', C=1000, max_iter=1000, solver=\"liblinear\") )\n",
    "]\n",
    ")\n",
    "\n",
    "titanic_logit2 = Logit2.fit(X_train, y_train)\n",
    "titanic_pred2 = Logit2.predict(X_train)\n",
    "print(classification_report(y_train, titanic_pred2))\n",
    "# accuracy 1\n",
    "print(np.mean(cross_val_score(Logit2, X_train, y_train, cv=10)))\n",
    "# ~0.84\n",
    "print(classification_report(y_test, Logit2.predict(X_test)))\n",
    "# ~0.85\n",
    "\n",
    "'''\n",
    "Support Vector Machine\n",
    "'''\n",
    "from sklearn.svm import SVC\n",
    "\n",
    "svc = Pipeline(steps=[ \n",
    "    ('preprocess', total_pipeline),\n",
    "    ('svm', SVC(C=100, kernel=\"linear\"))\n",
    "])\n",
    "\n",
    "titan_svc = svc.fit(X_train, y_train)\n",
    "pred_svc = svc.predict(X_train)\n",
    "print(classification_report(y_train, pred_svc))\n",
    "## 0.98.  \n",
    "\n",
    "'''\n",
    "dist_svc= { \n",
    "    \"svm__C\": range(1, 1000, 100),\n",
    "    \"svm__kernel\": ['linear', 'poly', 'rbf', 'sigmoid', 'precomputed']\n",
    "}\n",
    "\n",
    "svc_chosen = RandomizedSearchCV(svc, param_distributions=dist_svc, n_iter=50, random_state=4686)\n",
    "svc_chosen.fit(X_train, y_train)\n",
    "\n",
    "# svc_chosen.best_params_\n",
    "#{'svm__kernel': 'linear', 'svm__C': 1}\n",
    "# It takes 57 minutes. n*m*log_2(m)\n",
    "'''\n",
    "svc_chosen = Pipeline(steps=[ \n",
    "    ('preprocess', total_pipeline),\n",
    "    ('svm', SVC(C=1, kernel=\"linear\"))\n",
    "])\n",
    "\n",
    "svc_chosen.fit(X_train, y_train)\n",
    "pred_svc_rs = svc_chosen.predict(X_train)\n",
    "print(classification_report(y_train, pred_svc_rs))\n",
    "# 1 \n",
    "print(np.mean(cross_val_score(svc_chosen, X_train, y_train, cv=5)))\n",
    "# 0.83826\n",
    "\n",
    "print(classification_report(y_test, svc_chosen.predict(X_test)))\n",
    "#0.84\n",
    "\n",
    "\n",
    "'''\n",
    "Random Forest\n",
    "'''\n",
    "\n",
    "random_forest = Pipeline(steps=[ \n",
    "    ('preprocess', total_pipeline),\n",
    "    ('rf', RandomForestClassifier(criterion=\"entropy\", min_samples_leaf=1, random_state=4686))\n",
    "])\n",
    "\n",
    "titan_rf = random_forest.fit(X_train, y_train)\n",
    "pred_rf = random_forest.predict(X_train)\n",
    "print(classification_report(y_train, pred_rf))\n",
    "#  1\n",
    "np.mean(cross_val_score(random_forest, X_train, y_train, cv=10))\n",
    "# ~0.826. \n",
    "# own test data.\n",
    "print(classification_report(y_test, random_forest.predict(X_test)))\n",
    "# 0.84\n",
    "\n",
    "dist_rf = {\n",
    "      'preprocess__num_scale__poly__degree': range(2,4),\n",
    "      'rf__criterion': ['gini','entropy'],\n",
    "      'rf__max_depth': np.arange(5, 10),\n",
    "      'rf__min_samples_leaf': np.arange(1, 15, 2)\n",
    "}\n",
    "\n",
    "rf_chosen = GridSearchCV(random_forest, param_grid=dist_rf, cv=10)\n",
    "rf_chosen.fit(X_train, y_train)\n",
    "print(classification_report(y_train, rf_chosen.predict(X_train)))\n",
    "#0.88\n",
    "'''\n",
    "{'preprocess__num_scale__poly__degree': 2,\n",
    " 'rf__criterion': 'gini',\n",
    " 'rf__max_depth': 9,\n",
    " 'rf__min_samples_leaf': 1}\n",
    "'''\n",
    "\n",
    "print(classification_report(y_test, rf_chosen.predict(X_test)))\n",
    "#0.80 Not as good as the default model above; thus not used to \n",
    "\n",
    "# K-Nearest Neighbours\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "\n",
    "knc = Pipeline(steps=[ \n",
    "    ('preprocess', total_pipeline),\n",
    "    ('knclf', KNeighborsClassifier())\n",
    "])\n",
    "\n",
    "knc.fit(X_train, y_train)\n",
    "pred_knc = knc.predict(X_train)\n",
    "print(classification_report(y_train, pred_knc))\n",
    "#0.73\n",
    "\n",
    "disc_knc = {\n",
    "     'preprocess__num_scale__poly__degree': range(2,4),\n",
    "    \"knclf__n_neighbors\": range(0,100,5),\n",
    "    \"knclf__weights\":['uniform','distance']\n",
    "}\n",
    "\n",
    "knc_grid = RandomizedSearchCV(knc, param_distributions=disc_knc, cv=50, random_state=4686)\n",
    "knc_grid.fit(X_train, y_train)\n",
    "# 'preprocess__num_scale__poly__degree': 3, 'knclf__weights': 'distance', 'knclf__n_neighbors': 60\n",
    "\n",
    "pred_knc_grid = knc_grid.predict(X_train)\n",
    "print(classification_report(y_train, pred_knc_grid))\n",
    "# 1\n",
    "\n",
    "np.mean(cross_val_score(knc_grid, X_train, y_train, cv=10), 0)\n",
    "# ~ 0.6. knc_grid much overfitting.\n",
    "\n",
    "print(classification_report(y_test, knc_grid.predict(X_test)))\n",
    "#0.59\n",
    "\n",
    "# Try ensemble algorithm such as Voting classifiers.\n",
    "\n",
    "from sklearn.ensemble import VotingClassifier\n",
    "\n",
    "# Use the accuracies for test data as the weights\n",
    "vote_clf = VotingClassifier(\n",
    "    estimators= [('lr', lr_clf), ('knc', knc_grid), ('svc', svc_chosen), ('rf', random_forest)],\n",
    "    voting='hard'\n",
    ")\n",
    "\n",
    "vote_clf.fit(X_train, y_train)\n",
    "pred_vote = vote_clf.predict(X_train)\n",
    "print(classification_report(y_train, pred_vote))\n",
    "# 1\n",
    "np.mean(cross_val_score(vote_clf, X_train, y_train, cv=10), 0)\n",
    "# 0.8292175486205338\n",
    "pred_vote_test = vote_clf.predict(X_test)\n",
    "print(classification_report(y_test, pred_vote_test))\n",
    "#0.83\n",
    "\n",
    "vote_clf2 = VotingClassifier(\n",
    "    estimators= [('lr', lr_clf), ('svc', svc_chosen), ('rf', random_forest)],\n",
    "    voting='hard'\n",
    ")\n",
    "\n",
    "vote_clf2.fit(X_train, y_train)\n",
    "pred_vote2 = vote_clf2.predict(X_train)\n",
    "print(classification_report(y_train, pred_vote2))\n",
    "#\n",
    "np.mean(cross_val_score(vote_clf2, X_train, y_train, cv=10), 0)\n",
    "#0.8442107643600181\n",
    "print(classification_report(y_test, vote_clf2.predict(X_test)))\n",
    "#0.84\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3ffa08b9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-07-29T12:17:12.901467Z",
     "iopub.status.busy": "2022-07-29T12:17:12.901049Z",
     "iopub.status.idle": "2022-07-29T12:17:13.039244Z",
     "shell.execute_reply": "2022-07-29T12:17:13.038116Z"
    },
    "papermill": {
     "duration": 0.148084,
     "end_time": "2022-07-29T12:17:13.042066",
     "exception": false,
     "start_time": "2022-07-29T12:17:12.893982",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "test = pandas.read_csv('/kaggle/input/titanic/test.csv')\n",
    "# pred_test = random_forest.predict(test)\n",
    "np.sum(test.isna(),0)\n",
    "# Error: \"Fare\" has a missing value, in contrast to X_train and X_test.\n",
    "\n",
    "surname = []\n",
    "title = []\n",
    "for txt in test['Name']:\n",
    "      txt1 = txt.split(',')[0]\n",
    "      txt2 = (txt.split(\",\")[1]).split(\".\")[0]\n",
    "      surname.append(txt1)\n",
    "      title.append(txt2)\n",
    "\n",
    "test['Surname'] = pandas.DataFrame(surname)\n",
    "test['Title'] = pandas.DataFrame(title, dtype='category')\n",
    "test[\"Age\"] = test[\"Age\"].fillna(test.groupby(['Sex','Title'])['Age'].transform('mean'))\n",
    "# Not all missing values can be filled by group-by means; some are missing.\n",
    "test['Age'] = test['Age'].fillna(test['Age'].mean())\n",
    "test['Fare'] = test['Fare'].fillna(test.groupby(['Pclass'], dropna=True)['Fare'].transform('mean'))\n",
    "\n",
    "test2 = test.drop(['Name'], axis=1)\n",
    "pred_test = vote_clf2.predict(test2)\n",
    "\n",
    "res1 = (test.copy())['PassengerId']\n",
    "result = (pandas.DataFrame(res1)).assign(Survived = pred_test)\n",
    "result.to_csv('submission.csv',index=False)\n",
    "\n",
    "## "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 5392.948189,
   "end_time": "2022-07-29T12:17:13.881593",
   "environment_variables": {},
   "exception": null,
   "input_path": "__notebook__.ipynb",
   "output_path": "__notebook__.ipynb",
   "parameters": {},
   "start_time": "2022-07-29T10:47:20.933404",
   "version": "2.3.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
